compiler/lib/linker.ml

```diff
@@ -396,20 +423,40 @@ let load_fragment ~filename f =
                 let named_values = find_named_value code in
                 Primitive.register name kind ka arity;
                 StringSet.iter Primitive.register_named_value named_values;
-                (if Hashtbl.mem provided name
-                then
-                  let _, ploc, weakdef = Hashtbl.find provided name in
-                  if not weakdef
+                let is_isomorphic_js =
+                  Config.Flag.(include_browser_apis () && include_node_apis ())
+                in
+                let is_updating =
+                  if Hashtbl.mem provided name
                   then
-                    warn
-                      "warning: overriding primitive %S\n  old: %s\n  new: %s@."
-                      name
-                      (loc ploc)
-                      (loc pi));
-                Hashtbl.add provided name (id, pi, weakdef);
-                Hashtbl.add provided_rev id (name, pi);
-                check_primitive ~name pi ~code ~requires;
-                Hashtbl.add code_pieces id (code, requires)))
+                    let _, ploc, weakdef, prev_target_env = Hashtbl.find provided name in
+                    if not weakdef
+                    then
+                      match prev_target_env, target_env, is_isomorphic_js with
+                      | a, b, _ when a == b ->
+                          warn
+                            "warning: overriding primitive %S\n  old: %s\n  new: %s@."
+                            name
+                            (loc ploc)
+                            (loc pi);
+                          true
+                      | _, target_env, true -> target_env == `Isomorphic
```

Comment:
```suggestion
                      | _, `Isomorphic, true -> true
```

Reply:
the suggestion is not actually equivalent. equivalence would be:

```ml
| _, `Isomorphic, true -> true
| _, `Nodejs, true -> false
| _, `Browser, true -> false
```

i can do either. let me know your preference.

Reply:
it confused me when i wrote it, it confused you when you read it. it's evident that it was too confusing. I took another pass at this, and i think it's much simpler now.

see: 580ef47315